### PR TITLE
#1683 - Application Tracker Scholastic Change (Part 2)

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Extracted the scholastic standing change type from the dynamic data and persisted as a separate column using a DB enum.
- Created an error card to display the scholastic standing as an error when the change type is `Student did not complete program`.
image
_Right now there is no further information about what must happen with the disbursement/COE cards nor what should actually happen with the assessment. `Student did not complete program` is the only scholastic standing change type that does not generates a new student assessment._

### CODE REVIEW

**Review by ann-aot — APPROVED** [review]
Looks good 👍 Good work @andrewsignori-aot

File: sources/packages/web/src/types/contracts/StudentScholasticStanding.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * program enrolled by the student was not completed as expected and the institution
+ * must report the change to the Ministry.
+ */
+export enum StudentScholasticStandingChangeType {
+  /**
+   * Student went from full-time to part-time studies (vice-versa) or started a different program.
+   * This scholastic standing change type is only used to finish the current application. The actual
+   * changing from full-time to part-time studies (vice-versa) is not part of the scholastic standing change.
+   */
+  ChangeInIntensity = "Change in intensity",
```
Comment (dheepak-aot) [inline sources/packages/web/src/types/contracts/StudentScholasticStanding.ts:12]: Not part of the PR. But curious to know more about this type of Scholastic standing change @andrewsignori-aot  @JasonCTang

File: sources/packages/backend/libs/sims-db/src/entities/application.model.ts [hunk 2]
```diff
@@ -333,6 +334,19 @@ export class Application extends RecordDataModel {
...
   submittedDate?: Date;
+  /**
+   * Scholastic standings changes associated with this application.
+   */
+  @OneToMany(
+    () => StudentScholasticStanding,
+    (studentScholasticStanding) => studentScholasticStanding.application,
+    {
+      eager: false,
+      cascade: false,
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/sims-db/src/entities/application.model.ts:345]: Not about this PR. cascade by default is false. should we mention explicitly? I know we are doing this in many places. Me be we can stop defining explicitly.

File: sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json [hunk 3]
```diff
@@ -3294,22 +3294,22 @@
           "values": [
             {
               "label": "Student went from full-time to part-time studies (vice-versa) or started a different program",
-              "value": "changeInIntensity",
+              "value": "Change in intensity",
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json:3297]: Can these values load dynamically from Vue by using the enum so that we can avoid duplication of the same enum here?

Comment (ann-aot) [reply sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json:3297]: i belive these are formio key (api value)

File: sources/packages/web/src/types/contracts/StudentScholasticStanding.ts (hunk 1 above)
Comment (ann-aot) [reply sources/packages/web/src/types/contracts/StudentScholasticStanding.ts:12]: its  student changing from full-time to part-time studies and (vice-versa)

File: sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json:3297]: My point was to check if it is possible to load these radio values dynamically from vue.
image

File: sources/packages/web/src/types/contracts/StudentScholasticStanding.ts (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/web/src/types/contracts/StudentScholasticStanding.ts:12]: Thanks. I got that from comment, but my curiosity was to know when a student want to change from full-time to part time what happens on end to end? what happens to their tuition remittance and other costs involved related to the offering.?

**Comment by dheepak-aot** [comment]
Nice work @andrewsignori-aot 👍Just one minor comment. Other comments are not related to PR.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts [hunk 4]
```diff
@@ -533,7 +533,7 @@ export class ApplicationStudentsController extends BaseController {
           userToken.studentId,
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:533]: There's a code smell about duplicate block of code. Are you going to work on that?

File: sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts [hunk 5]
```diff
@@ -81,3 +81,8 @@ export const OFFERING_BULK_UPLOAD_MAX_FILE_SIZE = 4194304;
  * Expected length of the award value code.
  */
 export const AWARD_VALUE_CODE_LENGTH = 4;
+
+/**
+ * Minium unsuccessful weeks.
+ */
+export const MINIMUM_UNSUCCESSFUL_WEEKS = 68;
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts:88]: Minimum unsuccessful weeks for what? Could it be more specific, please?

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts [hunk 6]
```diff
@@ -543,13 +543,15 @@ export class ApplicationStudentsController extends BaseController {
...
     const [firstDisbursement, secondDisbursement] = disbursements;
+    const [scholasticStandingChange] = application.studentScholasticStandings;
     return {
       applicationStatusUpdatedOn: application.applicationStatusUpdatedOn,
       pirStatus: application.pirStatus,
       firstCOEStatus: firstDisbursement?.coeStatus,
       secondCOEStatus: secondDisbursement?.coeStatus,
       exceptionStatus: application.applicationException?.exceptionStatus,
       appealStatus,
+      scholasticStandingChangeType: scholasticStandingChange?.changeType,
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:554]: are we returning only the first scholasticStanding from the list?

**Review by andrepestana-aot — COMMENTED** [review]
Great job. Left some comments.

File: sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts [hunk 7]
```diff
@@ -81,3 +81,8 @@ export const OFFERING_BULK_UPLOAD_MAX_FILE_SIZE = 4194304;
  * Expected length of the award value code.
  */
 export const AWARD_VALUE_CODE_LENGTH = 4;
+
+/**
+ * Minium unsuccessful weeks.
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts:86]: Minimum typo

File: sources/packages/backend/libs/sims-db/src/entities/application.model.ts (hunk 2 above)
Comment (guru-aot) [reply sources/packages/backend/libs/sims-db/src/entities/application.model.ts:345]: You are right, the default is false, but we have been defining the relations like this for quite sometime.

File: sources/packages/web/src/types/contracts/StudentScholasticStanding.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/web/src/types/contracts/StudentScholasticStanding.ts:12]: @dheepak-aot, from business perspective @JasonCTang can provide more details.
From a system perspective, depending on the type, it may impact the offering end date or its associated costs (see images below).
For any of those cases, we will clone the existing offering, change the end date and/or costs and execute the assessment again.
To be clear, independent of the description, the above-mentioned scholastic type is not about changing from full-time to part-time, it is just about reporting that the current application is being finished earlier.

image

image

File: sources/packages/backend/libs/sims-db/src/entities/application.model.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/sims-db/src/entities/application.model.ts:345]: Yes, I am aware, I just kept doing what we have been doing 😉

File: sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/forms/src/form-definitions/reportscholasticstandingchange.json:3297]: We can, but I am not sure if we should. If they were used as text and value I would say 100% we should, in the current scenario, I am not sure.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts (hunk 6 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:554]: Yes, in the same way as `pirStatus`, `appealStatus`, `exceptionStatus`.

File: sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/route-controllers/application/application.students.controller.ts:533]: Planning to do it in the upcoming PR. Maybe further refactoring in that area because the recent changes generate more possible flow and it is more complex to follow right now.

File: sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts:88]: The variable was just moved, not introduced in this PR, but I see the point and I will change it.

File: sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts]: Comments reviewed.

File: sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts (hunk 5 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/api/src/utilities/system-configurations-constants.ts]: Comments reviewed.

**Review by dheepak-aot — APPROVED** [review]
Thanks for answering the questions. LGTM 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Review by andrepestana-aot — APPROVED** [review]
Looks good!
